epix/detector_volumes.py

```diff
@@ -1,129 +1,222 @@
+        min_z: Inclusive lower z boundary
+        max_z: Exclusive upper z boundary
+        max_r: Exclusive radial boundary
     """
-    r = np.sqrt(x**2 + y**2) 
+    r = np.sqrt(x**2 + y**2)
     m = r < max_r
     m = m & (z < max_z)
     m = m & (z >= min_z)
     return m
 
-@numba.njit()
-def clyinder_tpc(x, y, z): 
-    return in_cylinder(x, y, z, z_cathode, z_gate_mesh, sensitive_volume_radius)
-tpc = SensitiveVolume('tpc', 1, clyinder_tpc)
-
-@numba.njit()
-def bc(x, y, z): 
-    return in_cylinder(x, y, z, z_bottom_pmts, z_cathode, sensitive_volume_radius)
-below_cathode = SensitiveVolume('below_cathode', 2, bc)
 
 def in_sensitive_volume(events, sensitive_volumes):
     """
     Function which identifies which events are inside sensitive volumes.
+
+    Further, tests if sensitive volumes overlap. Assigns volume id, and
+    xenon density to interactions.
+
+    Args:
+        events (ak.records): Awkward record of the interactions.
+        sensitive_volumes (dict): Dictionary of the different volumes
+            defined via the SensitiveVolume class.
+
+    Returns:
+        ak.array: Awkward array containing the event ids.
     """
     for ind, vol in enumerate(sensitive_volumes):
         res = ak.ArrayBuilder()
-        res = _inside_sens_vol(events['x'], 
-                               events['y'], 
-                               events['z'], 
-                               vol.roi, 
-                               vol.volume_id, 
+        res = _inside_sens_vol(events['x'],
+                               events['y'],
+                               events['z'],
+                               vol.roi,
+                               vol.volume_id,
+                               vol.xe_density,
+                               vol.create_S2,
                                res)
         if ind:
             # Now we add the other results, but first test if 
-            # volumes overlap.
-            m = ak.any((result > 0) & (res == vol.volume_id))
-            assert not m, (f'The volume {vol.name} is overlapping with'
-                        ' an other volume!')   
-            result = result + res.snapshot()
+            # volumes overlap. Only possible if we explicitly loop
+            # over everything. This reduces performance but adds layer of
+            # safety.
+            m = (result['vol_id'] > 0) & (res['vol_id'] == vol.volume_id)
+            if ak.any(m):
+                overlapping_id = result[m][0]
+                # Get volume name:
+                name = [vol.name for vol in sensitive_volumes if vol.volume_id == overlapping_id][0]
+                raise ValueError(f'The volume {vol.name} is overlapping with'
+                                 f' volume {name}!')
+            new_results = res.snapshot()
+            for field in result.fields():
+                # Workaround since we cannot sum up records-arrays anymore
+                result[field] = result[field] + new_results[field]
```

Comment:
Shouldn't all the `result` in these lines be `res`? Otherwise I am missing what you do here.